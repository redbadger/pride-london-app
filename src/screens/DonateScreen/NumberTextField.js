// @flow
import React from "react";
import { Platform, StyleSheet, TextInput, View } from "react-native";
import type { ViewStyleProp } from "react-native/Libraries/StyleSheet/StyleSheet";
import Text, { scaleFont, scaleWithFont } from "../../components/Text";
import {
  lightNavyBlueColor,
  transparent,
  mediumGreyColor,
  eucalyptusGreenColor
} from "../../constants/colors";
import getAccessibiltyLabel from "./accessibilityLabel";

type Props = {|
  label: string,
  placeholder: string,
  onFocus: () => void,
  onChangeText: string => void,
  returnKeyLabel?: string,
  returnKeyType?: "done" | "go" | "next" | "search" | "send",
  onSubmitEditing?: () => void,
  style?: ViewStyleProp,
  value: ?string
|};

type State = {|
  focused: boolean,
  selection: ?{ start: number, end: number }
|};

class DonateScreen extends React.PureComponent<Props, State> {
  state = { focused: false, selection: undefined };

  onFocus = () => {
    this.setState({ focused: true });
    this.props.onFocus();
  };

  onBlur = () => {
    this.setState({ focused: false });
  };

  onSelectionChange = (e: {
    nativeEvent: { selection: { start: number, end: number }, target: number }
  }) => {
    this.setState({ selection: e.nativeEvent.selection });
  };

  render() {
    const { label, placeholder, onFocus, style, ...otherProps } = this.props;
    const { focused, selection } = this.state;

    // Note on accessibility:
    // React Native does not support to mark up labels for input fields
    // properly. See for example: https://github.com/facebook/react-native/issues/14989
    // This is why we use a custom `accessibilityLabel`, which tries to mimic,
    // TalkBack/VoiceOver behaviour.

    return (
      <View
        accessible
        accessibilityLabel={getAccessibiltyLabel(
          label,
          placeholder,
          otherProps.value,
          focused,
          selection
        )}
        style={style}
      >
        <Text
          type="h4"
          color="lightNavyBlueColor"
          importantForAccessibility="no"
          style={styles.label}
        >
          {label}
        </Text>
        <TextInput
          keyboardType="numeric"
          onBlur={this.onBlur}
          onFocus={this.onFocus}
          onSelectionChange={this.onSelectionChange}
          placeholder={focused ? placeholder : null}
          placeholderTextColor={mediumGreyColor}
          style={[styles.input, focused && styles.inputFocused]}
          underlineColorAndroid={transparent}
          importantForAccessibility="no"
          {...otherProps}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  label: {
    marginBottom: 8
  },
  input: {
    borderColor: mediumGreyColor,
    borderWidth: 1,
    borderRadius: 4,
    height: scaleWithFont("h2", 50),
    paddingHorizontal: 14,
    fontFamily: "Poppins-SemiBold",
    fontSize: scaleFont("h2", 18),
    color: lightNavyBlueColor,
    ...Platform.select({
      android: {
        lineHeight: scaleFont("h2", 18),
        includeFontPadding: false,
        textAlignVertical: "bottom"
      }
    })
  },
  inputFocused: {
    borderColor: eucalyptusGreenColor
  }
});

export default DonateScreen;
