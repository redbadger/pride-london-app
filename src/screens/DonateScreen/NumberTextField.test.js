// @flow
import React from "react";
import { TextInput } from "react-native";
import { shallow } from "enzyme";
import NumberTextField from "./NumberTextField";

it("renders correctly", () => {
  const output = shallow(
    <NumberTextField
      label="A number"
      placeholder="0"
      onFocus={() => {}}
      onChangeText={() => {}}
      value="13.37"
    />
  );
  expect(output).toMatchSnapshot();
});

it("renders placeholder when focused", () => {
  const onFocus = jest.fn();
  const output = shallow(
    <NumberTextField
      label="A number"
      placeholder="0.00"
      onFocus={onFocus}
      onChangeText={() => {}}
      value={null}
    />
  );

  expect(output.find(TextInput).prop("placeholder")).toBe(null);
  output.find(TextInput).simulate("focus");
  expect(onFocus).toHaveBeenCalled();
  expect(output.find(TextInput).prop("placeholder")).toBe("0.00");
  output.find(TextInput).simulate("blur");
  expect(output.find(TextInput).prop("placeholder")).toBe(null);
});
