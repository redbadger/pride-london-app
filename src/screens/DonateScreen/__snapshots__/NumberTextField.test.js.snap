// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  accessibilityLabel="A number, 13.37, text field; double tap to edit"
  accessible={true}
>
  <Text
    color="lightNavyBlueColor"
    importantForAccessibility="no"
    markdown={false}
    markdownStyle={Object {}}
    style={
      Object {
        "marginBottom": 8,
      }
    }
    type="h4"
  >
    A number
  </Text>
  <TextInput
    allowFontScaling={true}
    importantForAccessibility="no"
    keyboardType="numeric"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onSelectionChange={[Function]}
    placeholder={null}
    placeholderTextColor="#cecece"
    style={
      Array [
        Object {
          "borderColor": "#cecece",
          "borderRadius": 4,
          "borderWidth": 1,
          "color": "#2d2f7f",
          "fontFamily": "Poppins-SemiBold",
          "fontSize": 18,
          "height": 50,
          "paddingHorizontal": 14,
        },
        false,
      ]
    }
    underlineColorAndroid="transparent"
    value="13.37"
  />
</Component>
`;
